File: Leetcode/260. Single Number III.cpp
class Solution {
public:
    vector<int> singleNumber(vector<int>& a) {
        int i,j,k,n,aXORb = 0,x,y;
        x = y = 0;
        n  = a.size();
        vector<int>ans;
        if(n==0)
        return ans;
        for(i=0;i<n;i++)
        aXORb ^= a[i];
        k = aXORb & (-aXORb);
        for(i=0;i<n;i++){
            if(a[i]&k)
            x ^= a[i];
            else
            y ^= a[i];
        }
        ans.push_back(x);
        ans.push_back(y);
        return ans;
    }
};

File: Leetcode/260. Single Number III.java
https://leetcode.com/problems/single-number-iii/
class Solution {
    private int findXOROfArray(int[] set){
        int ans = 0;
        for(int i=0;i<set.length;i++){
            ans ^= set[i];
        }
        return ans;
    }
    
    public int[] singleNumber(int[] nums) {
        int n = nums.length;
        int xor = findXOROfArray(nums);
        int rightmostSetBit = xor & (~xor) +1;
        int num1=0;
        int num2=0;
        for(int i=0;i<n;i++){
            if((nums[i]&rightmostSetBit) == 0){
                num1 ^= nums[i];
            }else{
                num2 ^= nums[i];
            }
        }
        return new int[]{num1,num2};
        
    }
}
